lib/Model/Importer/DebugInfo/DwarfImporter.cpp

```diff
@@ -912,13 +914,19 @@ class DwarfToModelConverter : public BinaryImporterHelper {
       // Collect array whose elements are zero-sized
       //
       for (const model::QualifiedType &QT : Type->edges()) {
-        auto IsArray = [](const model::Qualifier &Q) {
-          return Q.Kind() == model::QualifierKind::Array;
-        };
-        if (llvm::any_of(QT.Qualifiers(), IsArray)
-            and not QT.UnqualifiedType().get()->size()) {
-          ToDrop.insert(Type.get());
+        auto &Unqualified = QT.UnqualifiedType();
+        if (Unqualified.isValid()) {
+          std::optional<uint64_t> Size = Unqualified.get()->trySize(VH);
+          if (Size.has_value())
+            continue;
         }
+
+        // At this point only invalid types and types with no size remain.
+        auto Iterator = revng::find_last_if_not(QT.Qualifiers(),
+                                                model::Qualifier::isConst);
+        if (Iterator != QT.Qualifiers().rend())
+          if (Iterator->Kind() == model::QualifierKind::Array)
+            ToDrop.insert(Type.get());
```

Comment:
It was dropping some valid arrays dwarf recovered because it was looking for **any** array qualifier on a zero-sized type, e.g. an array of `void *` or function pointers would get dropped even though it's valid (it's just an array of pointers).

Reply:
Ah nice, be more explicit about this in the commit message.

Reply:
What do you think?
```
DwarfImporter: prevent removal of valid arrays

The algorithm was looking for any array qualifier on a zero-sized
type, which led to some valid types being discarded in the case
when there was a pointer qualified in-between.

Consider that `void a[100]` is not a valid type, while
`void *a[100]` is.
```
https://github.com/revng/revng/pull/232/commits/9f79e6c60a565a35a0fa839d59eacd07bacf9b77